ci/run.sh

```diff
@@ -0,0 +1,116 @@
+#!/bin/bash
+set -e
+set -x
+
+THIS_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+ROOT_DIR=${THIS_DIR}/../
+
+BUILD_DIR=$(mktemp -d)
+INSTALL_DIR=$(mktemp -d)
+TEST_DIR=$(mktemp -d)
+trap "rm -rf ${BUILD_DIR} ${INSTALL_DIR} ${TEST_DIR}" EXIT
+
+if [ -n "${USE_CLANG}" ]; then
+  export CC=clang
+  export CXX=clang++
+fi
+
+if [ -n "${USE_NINJA}" ]; then
+  CMAKE_GENERATOR="Ninja"
+  MAKE=ninja
+  MAKE_J=
+else
+  CMAKE_GENERATOR="Unix Makefiles"
+  MAKE=make
+  MAKE_J="-j$(nproc)"
+fi
+
+# configure
+pushd ${BUILD_DIR}
+cmake -G "${CMAKE_GENERATOR}" \
+  -DCMAKE_INSTALL_PREFIX=${INSTALL_DIR} \
+  ${ROOT_DIR}
+
+# build docs
+${MAKE} doc 2> /dev/null
+
+# build project
+${MAKE} ${MAKE_J}
+${MAKE} install
+echo "Installation size: `du -sh ${INSTALL_DIR}`"
+
+
+export PATH=${INSTALL_DIR}/bin:${PATH}
+
+# run the unit and integration tests on the installed binaries to catch any
+# linking issues.
+
+# integration tests
+it_names=(
+wal_writer
+histograms
+rpc
+rpc_recv_timeout
+wal
+clock_pro
+cr
+)
+
+it_dirs=(
```

Comment:
we have 2 array syntaxes in this PR for bash scripts. 

one is 

```
arr="
1
2
3
4
"

and the other is 

arr=(
1
2
3
4
)

```

Wonder if we should make it consistent. 

I know that for the first case is for image names which has symbols that need to be scaped 

i.e.: `ubuntu:23`

Also should the elements of the array align with a space in front so they don't look like empty vars. 

i.e.: 

```
array=(
  1
  2
)
vs.
array=(
1
2
3
)
```

Reply:
the array construct lets bash index into the array (the other only supports list iteration). using the array is a hack for iterating over 2-tuples in bash, since for each test test_runner.py needs 1) test exec name and 2) source dir (e.g.  `test[i], test_dir[i]`). i'm confident that we can eliminate the usage of arrays and make this much simpler, but i think it will take some hacking on the test_runner.py.

Reply:
re: debug/release. sure, supporting these shouldn't be an issue. i'll add them to the to-do list.

what is common_build.sh used for?

are you at the point where you want to be versioning your releases, or is it too early?

Reply:
.common_build.sh can be deleted i think. just gets the nprocs. 

```
https://github.com/senior7515/smf/blob/master/.common_build.sh
```

re: versioning releases - I think that the RPC is the only subsystem that could. 

I think however that you would be the second system using it. 

There was a database vendor that i think forked the source code, but that's about it. 

What are your thoughts? my gut is we wait for versioning when there is one more project using it.?

Reply:
yeh, i'd hold off on versioning if there isn't anyone referencing the project that isn't also hacking on it.